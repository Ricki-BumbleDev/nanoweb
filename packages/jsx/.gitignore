*.js.map
*.d.ts
*.js
!*.config.js
