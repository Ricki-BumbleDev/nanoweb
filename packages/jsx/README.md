# nanoweb


nanoweb is a minimal library for building server-centric multi page web applications that behave like SPAs.

- **HTML-over-the-wire**: Merges HTML responses instead of replacing the complete page.
- **Based on standards**: Doesn't reinvent the wheel. It uses [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) and [web components](https://developer.mozilla.org/en-US/docs/Web/Web_Components).
- **Minimal**: Currently 560 LOC and the API consists of **only 5 functions**.
- **Component-based**: Build encapsulated components, then compose them to make complex UIs.

nanoweb's ideas can be freely combined with other server-centric frameworks while avoiding a technology lock-in, which makes it a perfect fit for [micro-frontends](https://martinfowler.com/articles/micro-frontends.html).

Check out the full documentation at: [https://nanoweb.js.org/](https://nanoweb.js.org/)
