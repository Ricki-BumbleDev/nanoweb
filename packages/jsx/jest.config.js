module.exports = {
  testMatch: ['**/tests/**/*.tsx'],
  testEnvironment: 'node',
  verbose: true,
  preset: 'ts-jest',
};
