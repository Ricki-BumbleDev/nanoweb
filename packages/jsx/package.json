{
  "name": "@nanoweb/jsx",
  "version": "0.0.6",
  "description": "nanoweb is a minimal library for building server-centric multi page web applications that behave like SPAs.",
  "main": "index.js",
  "types": "index.d.ts",
  "directories": {
    "test": "tests"
  },
  "scripts": {
    "test": "../../node_modules/.bin/jest",
    "build": "npm run clean && ../../node_modules/.bin/tsc --build --force",
    "clean": "git clean -Xdf"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/tbo/nanoweb.git"
  },
  "author": "Thomas Bonk",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/tbo/nanoweb/issues"
  },
  "homepage": "https://nanoweb.js.org/",
  "dependencies": {
    "csstype": "3.1.1"
  }
}
