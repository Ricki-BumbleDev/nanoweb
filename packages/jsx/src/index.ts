///<reference path="./typings.d.ts" />
const rxUnescaped = new RegExp(/["'&<>]/);

export const render = async (element: any): Promise<string> => {
  const awaited = element instanceof Promise ? await element : element;
  if (awaited instanceof Template) {
    return '<!DOCTYPE html>' + (await resolve(awaited));
  }
  return element;
};

const resolve = async (elements: any): Promise<string> => {
  let result = '';
  if (Array.isArray(elements)) {
    for (const child of elements) {
      result += await resolve(child instanceof Promise ? await child : child);
    }
  } else if (elements || elements === 0) {
    return elements;
  }
  return result;
};

function escape(text: any): string {
  /* Much faster when there is no unescaped characters */
  if (typeof text !== 'string' || !rxUnescaped.test(text)) {
    return text;
  }

  let result = '';
  let escape = '';
  let start = 0;
  let i: number;
  for (i = 0; i < text.length; ++i) {
    switch (text.charCodeAt(i)) {
      case 34: // "
        escape = '&quot;';
        break;
      case 39: // '
        escape = '&#039;';
        break;
      case 38: // &
        escape = '&amp;';
        break;
      case 60: // <
        escape = '&lt;';
        break;
      case 62: // >
        escape = '&gt;';
        break;
      default:
        continue;
    }
    if (i > start) {
      result += text.slice(start, i);
    }
    result += escape;
    start = i + 1;
  }
  return result + text.slice(start, i);
}

export class Template extends Array {}

export type TemplateType = typeof Template;

export const unsafe = (value = '') => Template.from([value]);

const SELF_CLOSING = new Set([
  'area',
  'base',
  'br',
  'col',
  'embed',
  'hr',
  'img',
  'input',
  'link',
  'meta',
  'param',
  'source',
  'track',
  'wbr',
]);

export const jsx = (type: any, props: Record<string, any>): Template => {
  if (typeof type === 'function') {
    return type(props);
  } else if (typeof type === 'undefined') {
    if (props.children) {
      return Template.from(props.children);
    } else {
      return new Template();
    }
  }
  let propString = '';
  if (props) {
    for (const i in props) {
      if (props[i] != null) {
        propString += toString(i, props[i]);
      }
    }
  }
  if (props.children == null) {
    const template = new Template();
    if (SELF_CLOSING.has(type)) {
      template[0] = '<' + type + propString + '/>';
    } else {
      template[0] = '<' + type + propString + '></' + type + '>';
    }
    return template;
  }
  const children =
    Array.isArray(props?.children) && !(props.children instanceof Template) ? props.children : [props.children];
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (!(child instanceof Template)) {
      children[i] = escape(child);
    }
  }
  const template = new Template();
  const length = children.length;
  template[0] = '<' + type + propString + '>';
  for (let i = 0; i < length; i++) {
    template[i + 1] = children[i];
  }
  template[length + 1] = '</' + type + '>';
  return template;
};

export const jsxs = jsx;
export const Fragment = (props: Record<string, any>) => jsx(undefined, props);

const CAMEL_CASE_PATTERN = new RegExp('[a-z]+((d)|([A-Z0-9][a-z0-9]+))*([A-Z])?', 'g');

const toKebapCase = (value: string) => value.match(CAMEL_CASE_PATTERN)!.join('-');

const kebapCache: Record<string, string> = {};
const keyCache: Record<string, string> = {};

const toString = (key: string, value: any) => {
  if (value == null || key === 'children') {
    return '';
  } else if (key === 'class' && Array.isArray(value)) {
    value = value.flat().filter(Boolean).join(' ');
  } else if (key === 'style') {
    value = Object.entries(value)
      .map(entry => [kebapCache[entry[0]] || (kebapCache[entry[0]] = toKebapCase(entry[0])), entry[1]].join(':'))
      .join(';')
      .toLowerCase();
  } else if (typeof value === 'object') {
    value = JSON.stringify(value);
  } else if (typeof value === 'boolean') {
    return ` ${keyCache[key] || (keyCache[key] = escape(key.toLowerCase()))}`;
  }
  return ` ${keyCache[key] || (keyCache[key] = escape(key.toLowerCase()))}="${escape(value)}"`;
};
