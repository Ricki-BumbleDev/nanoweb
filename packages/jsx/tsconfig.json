{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "module": "CommonJS",
    "target": "esnext",
    "types": ["node", "jest"],
    "lib": ["esnext"],
    "rootDir": "./src/",
    "outDir": ".",
    "jsx": "react-jsx",
    "jsxImportSource": "@nanoweb/jsx",
    "paths": {
      "@nanoweb/jsx": ["./src/jsx-runtime"],
      "@nanoweb/jsx/jsx-runtime": ["./src/jsx-runtime"]
    }
  },
  "exclude": ["node_modules"],
  "include": ["./src/**/*"]
}
