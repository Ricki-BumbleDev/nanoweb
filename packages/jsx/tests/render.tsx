import { render } from '@nanoweb/jsx';

const matchSnapshot = async (getComponent: () => JSX.Element) => expect(await render(getComponent())).toMatchSnapshot();

const Custom = (props: any) => <span>{props.children}</span>;

const AsyncCustom = async (props: any) => {
  await new Promise(resolve => setTimeout(resolve, 10));
  return <span>{props.children}</span>;
};

describe('Render to string', () => {
  test('Simple static component', async () => {
    await matchSnapshot(() => (
      <html>
        <head></head>
        <body>content</body>
      </html>
    ));
  });

  test('async component', async () => {
    await matchSnapshot(async () => (
      <html>
        <head></head>
        <body>content</body>
      </html>
    ));
  });

  test('Simple nesting', async () => {
    await matchSnapshot(() => (
      <html>
        <head>
          <title>{'a title'}</title>
        </head>
        <body>
          {123}
          {undefined}
          <p>{null}test</p>
        </body>
      </html>
    ));
  });

  test('Component nesting', async () => {
    await matchSnapshot(() => (
      <html>
        <head></head>
        <body>{<div>123</div>}</body>
      </html>
    ));
  });

  test('Self closing tags', async () => {
    await matchSnapshot(() => (
      <html>
        <head>
          <meta name="test" />
        </head>
        <body>
          <br />
          <hr />
          <img src="" />
        </body>
      </html>
    ));
  });

  test('Escape dangerous strings', async () => {
    await matchSnapshot(() => (
      <html>
        <head></head>
        <body>{'<script>alert("hacked")</script>'}</body>
      </html>
    ));
  });

  test('Async netsting', async () => {
    await matchSnapshot(() => (
      <html>
        <head></head>
        <body class="main">
          {Promise.resolve('hello')}
          {Promise.resolve(['world'])}
          <p>
            {Promise.resolve(
              <span style={{ color: 'red' }}>
                some {'text'}
                {Promise.resolve('!')}
              </span>,
            )}
          </p>
        </body>
      </html>
    ));
  });

  test('Custom component', async () => {
    await matchSnapshot(() => (
      <html>
        <head></head>
        <body class="main">
          <Custom>abc</Custom>
          <p>
            <AsyncCustom>def</AsyncCustom>
          </p>
          {[<Custom>ghi</Custom>, <AsyncCustom>jkm</AsyncCustom>]}
        </body>
      </html>
    ));
  });
});
