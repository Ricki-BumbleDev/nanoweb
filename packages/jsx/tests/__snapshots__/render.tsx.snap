// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render to string Async netsting 1`] = `"<!DOCTYPE html><html><head></head><body class="main">helloworld<p><span style="color:red">some text!</span></p></body></html>"`;

exports[`Render to string Component nesting 1`] = `"<!DOCTYPE html><html><head></head><body><div>123</div></body></html>"`;

exports[`Render to string Custom component 1`] = `"<!DOCTYPE html><html><head></head><body class="main"><span>abc</span><p><span>def</span></p><span>ghi</span><span>jkm</span></body></html>"`;

exports[`Render to string Escape dangerous strings 1`] = `"<!DOCTYPE html><html><head></head><body>&lt;script&gt;alert(&quot;hacked&quot;)&lt;/script&gt;</body></html>"`;

exports[`Render to string Falsy values 1`] = `"<!DOCTYPE html><html><head><title>a title</title></head><body>0<p>test</p>0<span>0</span><span></span><span></span><span>0</span></body></html>"`;

exports[`Render to string Render boolean attributes 1`] = `"<!DOCTYPE html><input type="checkbox"/>"`;

exports[`Render to string Render boolean attributes 2`] = `"<!DOCTYPE html><input type="checkbox" checked/>"`;

exports[`Render to string Render boolean attributes 3`] = `"<!DOCTYPE html><input type="checkbox" checked/>"`;

exports[`Render to string Render className attribute as class 1`] = `"<!DOCTYPE html><html><head></head><body class="awesome">content</body></html>"`;

exports[`Render to string Render null alt attribute 1`] = `"<!DOCTYPE html><img src="./test.jpg" alt=""/>"`;

exports[`Render to string Render object styles 1`] = `"<!DOCTYPE html><html><head></head><body style="text-align:center;border:1px;z-index:2">content</body></html>"`;

exports[`Render to string Render string styles 1`] = `"<!DOCTYPE html><html><head></head><body style="text-align: center; border: 1px">content</body></html>"`;

exports[`Render to string Self closing tags 1`] = `"<!DOCTYPE html><html><head><meta name="test"/></head><body><br/><hr/><img src=""/></body></html>"`;

exports[`Render to string Simple nesting 1`] = `"<!DOCTYPE html><html><head><title>a title</title></head><body>123<p>test</p></body></html>"`;

exports[`Render to string Simple static component 1`] = `"<!DOCTYPE html><html><head></head><body>content</body></html>"`;

exports[`Render to string Unsafe 1`] = `"<!DOCTYPE html><html><head></head><body class="main"><div><script>alert("hacked")</script></div><div><script>alert("hacked")</script></div><div><script>alert("hacked")</script></div><script>alert("hacked")</script><script>alert("hacked")</script></body></html>"`;

exports[`Render to string async component 1`] = `"<!DOCTYPE html><html><head></head><body>content</body></html>"`;
